include/deal.II/fe/fe_nedelec_sz.h

```diff
@@ -0,0 +1,273 @@
+                       const typename Mapping<dim,dim>::InternalDataBase              &mapping_internal,
+                       const dealii::internal::FEValues::MappingRelatedData<dim, dim> &mapping_data,
+                       const typename FiniteElement<dim,dim>::InternalDataBase        &fedata,
+                       dealii::internal::FEValues::FiniteElementRelatedData<dim, dim> &data) const;
+
+  /**
+  * To be documented.
+  */
+  virtual void
+  fill_fe_subface_values (const typename Triangulation<dim,dim>::cell_iterator           &cell,
+                          const unsigned int                                              face_no,
+                          const unsigned int                                              sub_no,
+                          const Quadrature<dim-1>                                        &quadrature,
+                          const Mapping<dim,dim>                                         &mapping,
+                          const typename Mapping<dim,dim>::InternalDataBase              &mapping_internal,
+                          const dealii::internal::FEValues::MappingRelatedData<dim, dim> &mapping_data,
+                          const typename FiniteElement<dim,dim>::InternalDataBase        &fedata,
+                          dealii::internal::FEValues::FiniteElementRelatedData<dim, dim> &data) const;
+
+  /**
+  * Derived Internal data which is used to store precomputed data.
+  * E.g. Once dim is known we can compute sigma, lambda, and combinations.
+  */
+  class InternalData : public FiniteElement<dim,dim>::InternalDataBase
+  {
+  public:
+    /**
+    * Storage for shape functions on the reference element
+    * We only pre-compute those cell-based DoFs, as the edge-based
+    * dofs depend on the choice of cell.
+    */
+    mutable std::vector<std::vector<Tensor<1,dim> > > shape_values;
+
+    mutable std::vector<std::vector<DerivativeForm<1, dim, dim> > > shape_grads;
+
+    /**
+    * TODO: remove the redundant parts, will need to update the 2D case when this is removed.
+    * - sigma - these won't be reused.
+    * - lambda - won't be reused.
+    * - lambda_ipj = lambda[i] + lambda[j] - no need, can go straight to the edge_lambda_*.
+    * - edgeDoF_to_poly - should remove the use of polyspace and just use the 1d polynomial for everything.
+    *
+    * Sigma_imj = sigma_i - sigma_j gives a parameterisation of an edge
+    * connected by vertices i and j. The values, gradients, non-zero
+    * component and sign of the co-efficient of x, y or z are stored.
+    * The orientation of edges and faces can then be handled using sigma_imj.
+    */
+    std::vector<std::vector<std::vector<double> > > sigma_imj_values;
+    std::vector<std::vector<std::vector<double> > > sigma_imj_grads;
+
+    /**
+    * Storage for "standard" edges:
+    * On edge, E_m = [e_{1}^{m}, e_{2}^{m}].
+    * edge_sigma[m][q] = sigma[e2][q] - sigma[e1][q]
```

Comment:
Sine this appears many times, can you please explain the difference between sigma and lambda.